packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+
+                e.addAutocorrect(core::AutocorrectSuggestion{
+                    fmt::format("Replace `{}` with `{}`", constLit->cnst.data(ctx)->shortName(ctx), replacement),
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        pName.fullName.emplace_back(constLit->cnst);
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+    reverse(pName.fullName.begin(), pName.fullName.end());
+    if (!pName.fullName.empty()) {
+        // Foo::Bar => Foo_Bar_Package
+        auto mangledName = absl::StrCat(absl::StrJoin(pName.fullName, "_", NameFormatter(ctx)), "_Package");
+        pName.mangledName = ctx.state.enterNameConstant(mangledName);
+    } else {
+        // Not a valid name.
+        pName.mangledName = core::NameRef::noName();
+    }
+    return pName;
+}
+
+bool isReferenceToPackageSpec(core::Context ctx, ast::TreePtr &expr) {
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    return constLit != nullptr && constLit->cnst == core::Names::Constants::PackageSpec();
+}
+
+unique_ptr<ast::UnresolvedConstantLit> copyConstantLit(ast::UnresolvedConstantLit *lit) {
+    auto copy = lit->deepCopy();
+    // Cast from Expression to UnresolvedConstantLit.
+    auto copyUcl = ast::cast_tree<ast::UnresolvedConstantLit>(copy);
+    ENFORCE(copyUcl != nullptr);
+    unique_ptr<ast::UnresolvedConstantLit> rv(copyUcl);
+    copy.release();
+    return rv;
+}
+
+ast::TreePtr name2Expr(core::NameRef name, ast::TreePtr scope = ast::MK::EmptyTree()) {
+    return ast::MK::UnresolvedConstant(core::LocOffsets::none(), move(scope), name);
+}
+
+ast::TreePtr PackageName::getFullNameLiteral(core::LocOffsets loc) const {
+    ast::TreePtr name = ast::MK::EmptyTree();
+    for (auto namePart : fullName) {
+        name = name2Expr(namePart, move(name));
+    }
+    // Outer name should have the provided loc.
+    if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(name)) {
+        name = ast::MK::UnresolvedConstant(loc, move(lit->scope), lit->cnst);
+    }
+    return name;
+}
+
+ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameRef fun, ast::TreePtr &expr) {
+    auto target = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    if (target == nullptr) {
+        if (auto e = ctx.beginError(expr->loc, core::errors::Packager::InvalidImportOrExport)) {
+            e.setHeader("Argument to `{}` must be a constant", fun.show(ctx));
+        }
+    }
+    return target;
+}
+
+core::LocOffsets firstLineOfFile(core::Context ctx) {
+    auto file = ctx.file;
+    core::LocOffsets firstLine{0, 0};
+    // Line 0 is reserved.
+    if (file.data(ctx).lineCount() > 1) {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).lineBreaks()[1]);
+    } else {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).source().length());
+    }
+    return firstLine;
+}
```

Comment:
I don't think it's worthwhile to have this helper. You could just as easily make a LocOffset of `{0, 0}` and the error-printing mechanism will show the full line of context for the error anyways.

Reply:
It also lets us get rid of these static_cast (which is how I found this function in the first place: ⌘F'ing for `static_cast` after making my comment about `ast::isa_tree` below)

Reply:
Does `LocOffset{0, 0}` work in IDE too?

Reply:
Yep. Feel free to try it yourself. zero-width locs are how the autocorrect for `sig`s work:

https://sorbet.run/#%23%20typed%3A%20strict%0Aclass%20A%0A%20%20def%20foo%3B%20end%0Aend

both those autocorrects show a single `^` , but in fact that's just how it presents a zero-width loc that starts and ends before that character.

Reply:
Fixed and removed.